### PR TITLE
Adds implementation of writer configuration

### PR DESCRIPTION
**Pull request by desaikd**
*Issue #618:*

*Description of changes:*
Adds an implementation of a writer configuration that can be used to create a writer based on format and Ion version. This uses `LazyEncoder` to build the writer using different `Encoding`s  with different `build_writer` implementations.

*List of changes:*
* Adds a writer configuration implementation that supports configuring both binary and text writer
* Adds` build_writer()` with `LazyEncoder` which can be used to be build a writer based on given writer configuration
* Removes write_as and write_all_as methods from element. Instead now `Lazy*Writer` can be used to write a Rust value as Ion.
* Uses `WriteConfig` within `to_binary` and `to_text` methods of element API

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: src/element/mod.rs [hunk 1]
```diff
@@ -923,10 +832,15 @@ assert_eq!(element_before, element_after);
 # }
 ```
     "##]
+    #[cfg(feature = "experimental-writer")]
     pub fn to_text(&self, text_kind: TextKind) -> IonResult<String> {
```
[the remark this answers is not in the document]
Comment (desaikd) [reply src/element/mod.rs]: Keeping these `Element` APIs as it is using the raw text/binary writer.

File: src/element/mod.rs [hunk 2]
```diff
@@ -881,10 +785,15 @@ assert_eq!(element_before, element_after);
 # }
 ```
 "##]
+    #[cfg(feature = "experimental-writer")]
     pub fn to_binary(&self) -> IonResult<Vec<u8>> {
```
[the remark this answers is not in the document]
Comment (desaikd) [reply src/element/mod.rs]: Keeping these Element APIs as it is using the raw text/binary writer as the lazy writer doesn't yet add strings to symbol table.

File: src/element/mod.rs [hunk 3]
```diff
@@ -749,104 +751,6 @@ impl Element {
...
-// Read the Element back from the serialized form
-let element_after = Element::read_one(&buffer)?;
-
-// Confirm that no data was lost
-assert_eq!(element_before, element_after);
-# Ok(())
-# }
-```
-"##]
-    pub fn write_as<W: io::Write>(&self, format: Format, output: W) -> IonResult<()> {
```
[the remark this answers is not in the document]
Comment (desaikd) [reply src/element/mod.rs]: Added back `write_as` and `write_as_all` methods as it was previously.

File: src/lazy/encoder/mod.rs [hunk 4]
```diff
@@ -54,6 +57,9 @@ pub trait LazyEncoder<W: Write>: 'static + Sized + Debug + Clone + Copy {
     where
         W: 'a;
     type EExpressionWriter<'a>;
+
+    /// Build writer based on given writer configuration
+    fn build_writer<E: Encoding>(config: WriteConfig<E>, output: W) -> Self::Writer;
```
[the remark this answers is not in the document]
Comment (desaikd) [reply src/lazy/encoder/mod.rs]: Added build method on `LazyRawWriter`

**Comment by desaikd** [comment]
Summary of new changes since last review:
* Modifies element APIs to use raw text/bianry writers instead of lazy
writer
* Adds implementation of `WriteAsIon` for `Element` and implementation
of `WriteAsIonValue` for `Value`
* adds build API implementation for binary and text lazy writers
* Puts `WriteConfig` under `experimental-lazy-reader` as it uses
`Encoding`

File: src/element/writer.rs [hunk 5]
```diff
@@ -3,13 +3,76 @@
...
+use std::marker::PhantomData;
+#[cfg(feature = "experimental-lazy-reader")]
+use {
+    crate::lazy::encoder::{LazyEncoder, LazyRawWriter},
+    crate::lazy::encoding::{BinaryEncoding_1_0, Encoding, TextEncoding_1_0},
+};
+
+/// Writer configuration to provide format and Ion version details to writer through encoding
+/// This will be used to create a writer without specifying which writer methods to use
+#[cfg(feature = "experimental-lazy-reader")]
```
[the remark this answers is not in the document]
Comment (desaikd) [reply src/element/writer.rs:22]: https://github.com/amazon-ion/ion-rust/issues/698

File: src/lazy/encoder/annotate.rs [hunk 6]
```diff
@@ -58,16 +57,23 @@ where
...
-        annotations_writer: V,
-    ) -> IonResult<()> {
-        let value_writer = annotations_writer.write_annotations(self.annotations.iter())?;
+    fn write_as_ion<V: AnnotatableValueWriter>(&self, writer: V) -> IonResult<()> {
+        let value_writer = writer.with_annotations(self.annotations);
         self.value.write_as_ion_value(value_writer)
     }
 }
+
+impl<'annotations, T, A> WriteAsIon for &Annotated<'annotations, T, A>
```
[the remark this answers is not in the document]
Comment (desaikd) [reply src/lazy/encoder/annotate.rs]: This was part of merge from `new-binary-writer` and is being used in examples for `annotated_with`

File: src/lazy/encoder/write_as_ion.rs [hunk 7]
```diff
@@ -35,6 +37,19 @@ pub trait WriteAsIon {
...
+    fn write_as_ion<V: AnnotatableValueWriter>(&self, writer: V) -> IonResult<()> {
+        let annotations = self.annotations().iter().collect::<Vec<&Symbol>>();
+        if !annotations.is_empty() {
+            self.value()
+                .write_as_ion_value(writer.with_annotations(&annotations))
+        } else {
+            self.value()
+                .write_as_ion_value(writer.without_annotations())
+        }
+    }
```
Comment (zslayton) [inline src/lazy/encoder/write_as_ion.rs:50]: We should add a `pub(crate)` way for this to access the `Annotations`' existing `Vec` so we don't have to `collect()` a new one. For now, you can optimize the no-annotations case a bit by doing:
```rust
if self.annotations().is_empty() {
    self.value().write_as_ion_value(writer.without_annotations())
} else {
    self.value.write_as_ion_value(writer.with_annotations(&Vec::from_iter(annotations.iter()))
}
```

**Review by zslayton — APPROVED** [review]

**Pull request merged**